pkg/finality-grandpa/past_rounds.go

```diff
@@ -0,0 +1,364 @@
+) *roundCommitter[Hash, Number, Signature, ID, E] {
+	return &roundCommitter[Hash, Number, Signature, ID, E]{
+		commitTimer, commitReceiver, nil,
+	}
+}
+
+func (rc *roundCommitter[Hash, Number, Signature, ID, E]) importCommit(
+	votingRound votingRound[Hash, Number, Signature, ID, E], commit Commit[Hash, Number, Signature, ID],
+) (bool, error) {
+	// ignore commits for a block lower than we already finalized
+	if votingRound.finalized() != nil && commit.TargetNumber < votingRound.finalized().Number {
+		return true, nil
+	}
+
+	base, err := votingRound.checkAndImportFromCommit(commit)
+	if err != nil {
+		return false, err
+	}
+	if base == nil {
+		return true, nil
+	}
+
+	rc.lastCommit = &commit
+
+	return true, nil
+}
+
+func (rc *roundCommitter[Hash, Number, Signature, ID, E]) commit(
+	waker *waker,
+	votingRound votingRound[Hash, Number, Signature, ID, E],
+) (bool, *Commit[Hash, Number, Signature, ID], error) {
+	rc.importCommits.setWaker(waker)
+loop:
+	for {
+		select {
+		case commit, ok := <-rc.importCommits.channel():
+			if !ok {
+				break loop
+			}
+			imported, err := rc.importCommit(votingRound, commit)
+			if err != nil {
+				return true, nil, err
+			}
+			if !imported {
+				log.Trace("Ignoring invalid commit")
+			}
+		default:
+			break loop
+		}
+	}
+
+	rc.commitTimer.SetWaker(waker)
+	elapsed, err := rc.commitTimer.Elapsed()
+	if elapsed {
+		if err != nil {
+			return true, nil, err
+		}
+	} else {
+		return false, nil, nil
+	}
+
+	lastCommit := rc.lastCommit
+	rc.lastCommit = nil
+	finalized := votingRound.finalized()
+
+	switch {
+	case lastCommit == nil && finalized != nil:
+		return true, votingRound.finalizingCommit(), nil
+	case lastCommit != nil && finalized != nil && lastCommit.TargetNumber < finalized.Number:
+		return true, votingRound.finalizingCommit(), nil
+	default:
+		return true, nil, nil
+	}
+}
+
+// A stream for past rounds, which produces any commit messages from those
+// rounds and drives them to completion.
+type pastRounds[Hash constraints.Ordered, Number constraints.Unsigned, Signature comparable,
+	ID constraints.Ordered, E Environment[Hash, Number, Signature, ID],
+] struct {
+	pastRounds    []backgroundRound[Hash, Number, Signature, ID, E]
+	commitSenders map[uint64]chan Commit[Hash, Number, Signature, ID]
+}
+
+func newPastRounds[Hash constraints.Ordered, Number constraints.Unsigned, Signature comparable,
+	ID constraints.Ordered, E Environment[Hash, Number, Signature, ID]]() *pastRounds[Hash, Number, Signature, ID, E] {
+	return &pastRounds[Hash, Number, Signature, ID, E]{
+		commitSenders: make(map[uint64]chan Commit[Hash, Number, Signature, ID]),
+	}
+}
+
+// push an old voting round onto this stream.
+func (p *pastRounds[Hash, Number, Signature, ID, E]) Push(env E, round votingRound[Hash, Number, Signature, ID, E]) {
+	roundNumber := round.roundNumber()
+	// TODO: this is supposed to be an unbounded channel on the producer side.  Use buffered in p.commitSenders
```

Comment:
still valid? if so lets create an issue for it

Reply:
I did in #3510